<font size="6">Week 3 Assignment:</font>
<br>
<br>
<font size="4">Part 1</font>
<br>
<br>
Explore and cluster the neighborhoods in Toronto.

In [4]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import requests

cana = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
cana_page = requests.get(cana)

df_cana = pd.read_html(cana_page.content, header=0)[0]

#Drop 'Not assigned' under Borough
df_cana = df_cana[df_cana.Borough != 'Not assigned']

df_cana.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [11]:
#Replace 'Not assigned' values in table
df_cana.Neighborhood.replace('Not assigned', df_cana.Borough, inplace=True)
df_cana.head(10)

Postal Code           Borough                                 Neighborhood
2          M3A        North York                                    Parkwoods
3          M4A        North York                             Victoria Village
4          M5A  Downtown Toronto                    Regent Park, Harbourfront
5          M6A        North York             Lawrence Manor, Lawrence Heights
6          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8          M9A         Etobicoke      Islington Avenue, Humber Valley Village
9          M1B       Scarborough                               Malvern, Rouge
11         M3B        North York                                    Don Mills
12         M4B         East York              Parkview Hill, Woodbine Gardens
13         M5B  Downtown Toronto                     Garden District, Ryerson

In [18]:
# Group neighbourhoods with same Postal code
df_canada = df_cana.groupby(['Postal Code', 'Borough'])['Neighborhood'].apply(lambda x: ', '.join(x))
df_canada = df_canada.reset_index()
df_canada.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [19]:
# Print the shape
df_canada.shape


(103, 3)

As seen above, 103 rows are printed out (even though original dataframe has 212 rows). Some rows may have the same postal code more than once, and we have merged them into the same row. 3 columns are Postal Code, Borough and Neighborhood. Hence the final result. 

<font size="4">Part 2</font>
<br>
<br>
Get the latitude and the longitude coordinates of each neighborhood

In [22]:
# Import csv file


lonlat = pd.read_csv('https://cocl.us/Geospatial_data')
lonlat.head()


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [25]:
df_cantor = pd.merge(df_canada, lonlat, on='Postal Code')
df_cantor.head(20)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   
15         M1W  Scarborough   
16         M1X  Scarborough   
17         M2H   North York   
18         M2J   North York   
19         M2K   North York   

                                         Neighborhood   Latitude  Longitude  
0                                      Malvern, Rouge  43.806686 -79.194353  
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029  
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577  
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476  
9                         Birch Cliff, Cliffside West  43.692657 -79.264848  
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304  
11                                  Wexford, Maryvale  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302  
14  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577  
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389  
16                                        Upper Rouge  43.836125 -79.205636  
17                                  Hillcrest Village  43.803762 -79.363452  
18                       Fairview, Henry Farm, Oriole  43.778517 -79.346556  
19                                    Bayview Village  43.786947 -79.385975

Instruction: Explore and cluster the neighborhoods in Downtow, East, Central & West Toronto. 

In [69]:
df_toronto = df_cantor[df_cantor['Borough'].str.contains('Downtown Toronto|East Toronto|West Toronto|Central Toronto')]
df_toronto

Postal Code           Borough  \
37         M4E      East Toronto   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
87         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                  North Toronto West, Lawrence Park  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
66                     University of Toronto, Harbord  43.662696 -79.400049  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
69                                     Stn A PO Boxes  43.646435 -79.374846  
70             First Canadian Place, Underground city  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564  
76                       Dufferin, Dovercourt Village  43.669005 -

In [31]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium 

# import folium to visualize clusters
print('Folium installed')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    ------------------------------------------------------------
                       

In [70]:
map_toronto = folium.Map(location=[43.651070,-79.347015],zoom_start=12)

for lat, lon, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lon],
    radius=5,
    popup=label,
    color='blue',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_toronto)
    
map_toronto

<font size="4">PS: The map might not be visible on Github. Please run the programme on Jupyter Notebook or Watson Studio to visualize.</font>

In [83]:
# import k-means for clustering
from sklearn.cluster import KMeans
from sklearn.datasets.samples_generator import make_blobs

In [85]:
k=5
toronto_cluster = df_toronto.drop(['Postal Code','Borough','Neighborhood'],1)
kmeans = KMeans(n_clusters = k,random_state=0).fit(toronto_cluster)
kmeans.labels_[0:10]


array([0, 0, 0, 0, 1, 1, 1, 1, 1, 1], dtype=int32)

In [87]:

df_toronto

Cluster Labels Postal Code           Borough  \
37               1         M4E      East Toronto   
41               1         M4K      East Toronto   
42               1         M4L      East Toronto   
43               1         M4M      East Toronto   
44               0         M4N   Central Toronto   
45               0         M4P   Central Toronto   
46               0         M4R   Central Toronto   
47               0         M4S   Central Toronto   
48               0         M4T   Central Toronto   
49               0         M4V   Central Toronto   
50               1         M4W  Downtown Toronto   
51               1         M4X  Downtown Toronto   
52               1         M4Y  Downtown Toronto   
53               1         M5A  Downtown Toronto   
54               1         M5B  Downtown Toronto   
55               1         M5C  Downtown Toronto   
56               1         M5E  Downtown Toronto   
57               1         M5G  Downtown Toronto   
58               1         M5H  Downtown Toronto   
59               1         M5J  Downtown Toronto   
60               1         M5K  Downtown Toronto   
61               1         M5L  Downtown Toronto   
63               0         M5N   Central Toronto   
64               0         M5P   Central Toronto   
65               0         M5R   Central Toronto   
66               1         M5S  Downtown Toronto   
67               1         M5T  Downtown Toronto   
68               1         M5V  Downtown Toronto   
69               1         M5W  Downtown Toronto   
70               1         M5X  Downtown Toronto   
75               2         M6G  Downtown Toronto   
76               2         M6H      West Toronto   
77               2         M6J      West Toronto   
78               2         M6K      West Toronto   
82               2         M6P      West Toronto   
83               2         M6R      West Toronto   
84               2         M6S      West Toronto   
85               1         M7A  Downtown Toronto   
87               1         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                  North Toronto West, Lawrence Park  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64    Forest Hill North & West, Forest Hill Road Pa

In [91]:
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors

#import libraries for clustering

In [97]:
map_clusters = folium.Map(location=[43.651070,-79.347015],zoom_start=12)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, neighbourhood, cluster in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Neighborhood'], df_toronto['Cluster Labels']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters